### PR TITLE
Fixes for exporting labels/text as text objects, not paths

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR is a bunch of ground work to finally and permanently closing off the long standing https://issues.qgis.org/issues/3975 issue.

In order to properly solve this issue and make sure that the user's desire to either export ALL text (not just labels -- this also applies to things like scale bar text now) is respected EVERYWHERE I've moved the setting for text renderer type into QgsRenderContext. This allows the setting to be controlled on a render by render basis, so it fixes the fragile/buggy layout export override for this setting which previously relied on temporarily changing a project-wide setting.

In order to future proof this I've used an enum for the text format. The current options are `TextFormatAlwaysOutlines` (default) and `TextFormatAlwaysText`. The intention here is that we could potentially add a third option `PreferText` -- which means "use text objects, unless doing so results in inferior rendering"** (and make this the new default).

For consistency and ease of use I've also duplicated the existing dialog which shows on exporting a layout to SVG where users can set export options such as the text format and added it for PDF export. This allows us to fix #8844 and let users easily control text settings at PDF export time. It also brings more consistency -- an export options dialog is now shown for all export types from layouts (pdf/svg/raster).

The final part left is testing that older workarounds around OSX PDF export issues can now be removed due to Qt issues fixed in Qt5 -- @3nids will handle this testing and fix.

* Interestingly, trawling through the archives here and testing on master leads me to believe that the issues which caused us to move from text -> paths as defaults back in QGIS 2.0 have been fixed in Qt. At least, I can't find any regressions in the output quality when using text now (including previously mentioned issues like labels with a buffer/underline combination). But this discussion/change can wait for a future PR/release. (ping @dakcarto )

### CODE REVIEW

**Comment by nyalldawson** [comment]
I should have also added: the approach used here makes the text export "just work" using the correct setting whenever QgsTextRenderer is used. This also fixes a regression in 3.2 where scalebar text is always forced to paths. Now scalebar text, along with map labels, will also correctly respect the text output format and export as text when the user has picked this export option.

It allows us to safely use QgsTextRenderer everywhere without any risk of regressions and be confident that the export setting will work automatically.

**Comment by nirvn** [comment]
Tested the branch, works nicely.

On the UI front, I think we could try and harmonize things a bit. Putting the three export option dialogs next to each other:
image

- either rename the 'SVG Options' and 'PDF Options' collapsible groups to 'Export Options', or rename the latter to 'Image Optins'
- we need to have a consistent use of bold font for the collapsible group titles
- for the image export, IMHO we need to move the [x] Generate world file and [x] Enable antialiasing checkboxes _into_ the Export/Image Options collapsible group.

**Comment by nirvn** [comment]
Also, `top margin` vs `top margin (mm)` needs to be harmonized.

File: src/app/layout/qgslayoutdesignerdialog.cpp [hunk 1]
```diff
@@ -2064,12 +2066,15 @@ void QgsLayoutDesignerDialog::exportToPdf()
 
   mView->setPaintingEnabled( false );
   QgsTemporaryCursorOverride cursorOverride( Qt::BusyCursor );
+
+  QgsLayoutExporter::PdfExportSettings pdfSettings;
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/app/layout/qgslayoutdesignerdialog.cpp]: Thanks, fixed

**Comment by nyalldawson** [comment]
Thanks @nirvn, all done

**Comment by 3nids** [comment]
I just tested this under mac and text is searchable and selectable. Is there anything more I should test?

**Comment by nyalldawson** [comment]
@3nids is that with the native format change too?

**Comment by 3nids** [comment]
what do you mean the native format change?

**Comment by nyalldawson** [comment]
@wonder-sk do you mind quickly looking over this new API?

**Comment by 3nids** [comment]
@nyalldawson yes, with the native change, texts are selectable and searchable!

**Comment by 3nids** [comment]
although, it is indeed always exporting to portrait letter.
but I don't see any reason to use Native format? it looks like the issue of font embedding has been fixed.

**Comment by nyalldawson** [comment]
@3nids in that case, I believe all that's left to do here is remove the outdated comment?

I was also looking at that block:

      // TODO: add option for this in layout
      // May not work on Windows or non-X11 Linux. Works fine on Mac using QPrinter::NativeFormat
      //printer.setFontEmbeddingEnabled( true );

and trying to work out if we should expose this as an option. But Qt docs don't have any detail on side effects of this call, and the wider web doesn't have much info either. So I'm inclined to just leave this as a comment/TODO for now

**Comment by wonder-sk** [comment]
@nyalldawson the API looks good to me. (I have not managed to properly review the bulk of the code, sorry)

If I may suggest one thing though, the text format options could use some extra documentation. Not having played a lot with these options I struggle to understand all implications. For example:
- when rendering to raster image, do I care about these options? Is any of those superior? (I would expect Text may give better result as it could in theory use hinting of the OS for less blurry appearance with small font sizes?)
- when rendering to PDF with Text, do the fonts get (optionally?) stored in the PDF so don't need to worry about getting wrong rendering on a different computer?
- when rendering to SVG with Text, I guess the fonts always need to be shipped separately, right? I see SVG has some support for embedding of fonts and their glyphs (https://www.w3.org/TR/SVGTiny12/fonts.html), but this is probably not supported by QtSvg?
- for Outlines format, it may be worth explicitly noting that since the text gets converted to curves, the downside is that the text content can't be edited by third-party software
- for Text the docstring says - "This may result in rendering artifacts or poor quality rendering" - when does this happen? is it platform specific?

**Comment by nyalldawson** [comment]
Thanks @wonder-sk !

(Disclaimer: I'm going off bug reports and discussion from ~>4 years ago, so some/all may no longer apply. But it's hard to find Qt docs on this in general)

> when rendering to raster image, do I care about these options? Is any of those superior? (I would expect Text may give better result as it could in theory use hinting of the OS for less blurry appearance with small font sizes?)

Actually, past experience is the opposite. Text as curves gives better raster output, which is why this is default behaviour for canvas too. 

the-grinch-poster1-700x300


Refs https://github.com/qgis/QGIS/pull/1286#issuecomment-39806854, my personal experience when flipping default behaviour to text as text and seeing the rendering differences in the unit tests.

> when rendering to PDF with Text, do the fonts get (optionally?) stored in the PDF so don't need to worry about getting wrong rendering on a different computer?

Refs https://github.com/qgis/QGIS/blob/master/src/core/layout/qgslayoutexporter.cpp#L1031. But that comment appears outdated. I actually did have a commit in this branch exposing this option to users, but when I tested it appears as though (on Linux, at least), Qt pays no attention to the QPrinter::setFontEmbeddingEnabled setting and ALWAYS embeds fonts. There's no clues I can find on the web about expected behaviour of this flag or side effects, so I deleted that commit as a "come back to later if there's demand" thing.

Behaviour may be different on different platforms. Qt... Woot.

1ca76ce599a8b9c975dbdcfebf02c085_700x259


> when rendering to SVG with Text, I guess the fonts always need to be shipped separately, right? 

Correct. Apps like inkscape/illustrator warn when you open SVGs and they contain missing fonts.

> but this is probably not supported by QtSvg

Pft. Past experience shows that Qt upstream cares little about SVG and the multitude of serious open bugs it's SVG generator/renderer has. So no chance of this being supported by QtSvg.


grinch1


> for Outlines format, it may be worth explicitly noting that since the text gets converted to curves, the downside is that the text content can't be edited by third-party software

Sure. Just to clarify, are you referring to the dox or within the export settings UI?

> for Text the docstring says - "This may result in rendering artifacts or poor quality rendering" - when does this happen? is it platform specific?

Refs https://github.com/qgis/QGIS/pull/1286. Quoting @dakcarto "Since the label buffer is already drawn as text-as-outline, the use of text-as-text over the buffer resulted in poor registration. Text-as-text does not include any allowance for stroke width, which leads to some styling features, e.g. underline, to be offset from the buffer with no easy means of fixing it (Qt glyph and style rendering).". 

I tried to reproduce this on Qt 5 and didn't see any issues with the buffer + underline combination, so it's possible that this issue at least is fixed. But I'm not comfortable changing the default setting here without extensive testing. There's also the above described rendering degradation when using text as text on raster paint surfaces.

I'd love for someone to conduct and in-depth, cross-platform test to make a list of settings which cause text-as-text issues, so that we can use these and add a "prefer text-as-text" setting which only falls back to curves when these settings are present. But that's a separate piece of work, and would rely on the move to the setting enum made here.

**Comment by wonder-sk** [comment]
Thanks a lot @nyalldawson for the elaborate summary and beautiful illustrations of a poor developer's feelings :)

Having a short version of this in API docs would be extremely useful so we do not need to investigate again all the related issues.

> Just to clarify, are you referring to the dox or within the export settings UI?

I meant just dox... probably too much detail for an ordinary user :)

**Comment by nyalldawson** [comment]
Thanks @wonder-sk, dox expanded

**Comment by 3nids** [comment]
@nyalldawson great job! do you want to wait before backporting?

**Comment by nyalldawson** [comment]
@3nids it's already backported.

**Pull request merged**